checker/src/org/checkerframework/checker/index/upperbound/UpperBoundAnnotatedTypeFactory.java

```diff
@@ -380,10 +396,56 @@ public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror
         @Override
         public Void visitUnary(UnaryTree node, AnnotatedTypeMirror type) {
             // Dataflow refines this type if possible
-            type.addAnnotation(UNKNOWN);
+            if (node.getKind() == Kind.BITWISE_COMPLEMENT) {
+                addAnnotationForBitwiseComplement(
+                        getSearchIndexAnnotatedTypeFactory().getAnnotatedType(node.getExpression()),
+                        type);
+            } else {
+                type.addAnnotation(UNKNOWN);
+            }
             return super.visitUnary(node, type);
         }
 
+        /**
+         * If a type returned by an {@link SearchIndexAnnotatedTypeFactory} has a {@link
+         * NegativeIndexFor} annotation, then refine the result to be {@link LTEqLengthOf}. This
+         * handles this case:
+         *
+         * <pre>{@code
+         * int i = Arrays.binarySearch(a, x);
+         * if (i >= 0) {
+         *     // do something
+         * } else {
+         *     i = ~i;
+         *     // i is now @LTEqLengthOf("a"), because the bitwise complement of a NegativeIndexFor is an LTL.
+         *     for (int j = 0; j < i; j++) {
+         *          // j is now a valid index for "a"
+         *     }
+         * }
+         * }</pre>
+         *
+         * @param searchIndexType The type of an expression in a bitwise complement. For instance,
+         *     in {@code ~x}, this is the type of {@code x}.
+         * @param typeDst The type of the entire bitwise complement expression. It is modified by
+         *     this method.
+         */
+        private void addAnnotationForBitwiseComplement(
+                AnnotatedTypeMirror searchIndexType, AnnotatedTypeMirror typeDst) {
+            if (AnnotationUtils.containsSameByClass(
+                    searchIndexType.getAnnotations(), NegativeIndexFor.class)) {
+                AnnotationMirror nif = searchIndexType.getAnnotation(NegativeIndexFor.class);
+                List<String> arrays = IndexUtil.getValueOfAnnotationWithStringArgument(nif);
+                List<String> negativeOnes = new ArrayList<>(arrays.size());
+                for (String s : arrays) {
+                    negativeOnes.add("-1");
```

Comment:
You can use 
` List<String> negativeOnes = Collections.nCopies(arrays.size(), "-1");`
instead of this for loop.